src/v2/guide/typescript.md

```diff
@@ -139,3 +139,38 @@ var vm = new Vue({
   myOption: 'Hello'
 })
 ```
+
+## Annotating Return Type
+
+Vue's type definition is prone to [circular reference](https://github.com/Microsoft/TypeScript/issues/12846#issuecomment-270296195) of `this`.
+Annotating return type is highly recommended to avoid such error.
+
+```ts
+import Vue, { VNode } from 'vue'
+
+const Component = Vue.extend({
+  data() {
+    return {
+      msg: 'Hello'
+    }
+  },
+  methods: {
+    // need annotation due to `this` in return type
+    greet(): string {
+      return this.msg + ' world'
+    }
+  },
+  computed: {
+    // need annotation
+    greeting(): string {
+      return this.greet() + '!'
+    }
+  },
+  // `createElement` is inferred, but `render` needs return type
```

Comment:
I believe the reason you're running into this is that it references `greeting`. You shouldn't need this if `greeting` is annotated.

Reply:
I have tried it but sadly `render` still needs annotation. (as reported in vuejs/vue#6921).

I guess the reason is that compiler needs to check the call of `createElement` to get the return type, and check calling will cause check argument types. In this case, `this.greeting` causes cyclic error.